runtime/src/main/cpp/Memory.cpp

```diff
@@ -635,13 +637,11 @@ void MarkGray(ContainerHeader* start) {
     auto container = toVisit.front();
     toVisit.pop_front();
     if (useColor) {
-      if (container->color() == CONTAINER_TAG_GC_GRAY) continue;
-    } else {
-       if (container->marked()) continue;
-    }
-    if (useColor) {
+      int color = container->color();
+      if (color == CONTAINER_TAG_GC_GRAY || color == CONTAINER_TAG_GC_GREEN) continue;
       container->setColor(CONTAINER_TAG_GC_GRAY);
     } else {
+      if (container->marked()) continue;
```

Comment:
Do we have to traverse green objects here?

Reply:
Seems we shall.